src/log-store/src/kafka/util/record.rs

```diff
@@ -0,0 +1,608 @@
+impl TryFrom<Record> for KafkaRecord {
+    type Error = crate::error::Error;
+
+    fn try_from(record: Record) -> Result<Self> {
+        let key = serde_json::to_vec(&record.meta).context(EncodeJsonSnafu)?;
+        Ok(KafkaRecord {
+            key: Some(key),
+            value: Some(record.data),
+            timestamp: rskafka::chrono::Utc::now(),
+            headers: Default::default(),
+        })
+    }
+}
+
+impl TryFrom<KafkaRecord> for Record {
+    type Error = crate::error::Error;
+
+    fn try_from(kafka_record: KafkaRecord) -> Result<Self> {
+        let key = kafka_record.key.context(MissingKeySnafu)?;
+        let meta = serde_json::from_slice(&key).context(DecodeJsonSnafu)?;
+        let data = kafka_record.value.context(MissingValueSnafu)?;
+        Ok(Self { meta, data })
+    }
+}
+
+impl From<Vec<Record>> for EntryImpl {
+    fn from(records: Vec<Record>) -> Self {
+        let entry_id = records[0].meta.entry_id;
+        let ns = records[0].meta.ns.clone();
+        let data = records.into_iter().flat_map(|record| record.data).collect();
+        EntryImpl {
+            data,
+            id: entry_id,
+            ns,
+        }
+    }
+}
+
+/// Produces a record to a kafka topic.
+pub(crate) struct RecordProducer {
+    /// The namespace of the entries.
+    ns: NamespaceImpl,
+    /// Entries are buffered before being built into a record.
+    entries: Vec<EntryImpl>,
+}
+
+impl RecordProducer {
+    /// Creates a new producer for producing entries with the given namespace.
+    pub(crate) fn new(ns: NamespaceImpl) -> Self {
+        Self {
+            ns,
+            entries: Vec::new(),
+        }
+    }
+
+    /// Populates the entry buffer with the given entries.
+    pub(crate) fn with_entries(self, entries: Vec<EntryImpl>) -> Self {
+        Self { entries, ..self }
+    }
+
+    /// Pushes an entry into the entry buffer.
+    pub(crate) fn push(&mut self, entry: EntryImpl) {
+        self.entries.push(entry);
+    }
+
+    /// Produces the buffered entries to Kafka sever. Those entries may span several Kafka records.
+    pub(crate) async fn produce(self, client_manager: &ClientManagerRef) -> Result<Offset> {
+        ensure!(!self.entries.is_empty(), EmptyEntriesSnafu);
+
+        // Gets the producer in which a record buffer is maintained.
+        let producer = client_manager
+            .get_or_insert(&self.ns.topic)
+            .await
+            .map_err(|e| {
+                GetClientSnafu {
+                    topic: &self.ns.topic,
+                    error: e.to_string(),
+                }
+                .build()
+            })?
+            .producer;
+
+        // Stores the offset of the last successfully produced record.
+        let mut last_offset = None;
+        let max_record_size =
+            client_manager.config.max_batch_size.as_bytes() as usize - ESTIMATED_META_SIZE;
+        for entry in self.entries {
+            for record in build_records(entry, max_record_size) {
```

Comment:
Does rskafka support transaction? If no, them some records may be failed to produce to kafka, making the entry incomplete to replay.

Reply:
That's a critical problem!

Reply:
Created an issue for the problem, see: https://github.com/GreptimeTeam/greptimedb/issues/3108

Reply:
If it fails to produce a record, it throws an error to the upper(finally reaching the end-user). It seems it's acceptable. While replaying, the system will ignore broken records(i.e., missing parts).